spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractMessageListenerContainer.java

```diff
@@ -277,6 +277,10 @@ public boolean isRunning() {
 		return this.running;
 	}
 
+	protected boolean canStop() {
+		return this.running;
```

Comment:
Why `isRunning()` not enough?
It is totally OK to have an extra logic in the overridden method in that `ConcurrentMessageListenerContainer`.

Reply:
Assume concurrency == 2.
Cmain -- concurrent container
C0, C1 -- child containers
   
1. `start` the concurrent container. 

   Cmain -- running
   C0 -- running
   C1 -- running
2. `stop` container C0 manually.
   
   Cmain -- not running
   C0 -- not running
   C1 -- running
   
3. `stop` 'Cmain' container. 
   
   As per the earlier condition concurrent container running status is false. So, it will not be stopped. 
   
   New condition has to be added to verify if really child containers are cleared. This is equal to `stop` is called prior or not.

Reply:
> stop container C0 manually.

> Cmain -- not running
> C0 -- not running
> C1 -- running

But this situation is not correct.
The `Cmain` has to be running until any of its children is running.

Reply:
But as per the API definition,

From interface:
org.springframework.context.Lifecycle Check whether this component is currently running.
In the case of a container, this will return true only if all components that apply are currently running

Reply:
> The `Cmain` has to be running until any of its children is running.

`Cmain` ConcurrentContainer is not stopped. Only it's running status will be set to `false` to indicate that one or all containers are stopped as per the API definition.

Reply:
> > stop container C0 manually.
> 
> > Cmain -- not running
> > C0 -- not running
> > C1 -- running
> 
> But this situation is not correct. The `Cmain` has to be running until any of its children is running.

It could be defined like this.

1) set `running` true if all the contianers are running. Indicates `green` and All is Well.
2) set `running` false if any or all the containers are not running. Indicates `red` and all is not well. In this case, App dev can iterate over the containers to verify what are all stopped.
3) `isChildRunningAPI` returns `false` only when all the message processing is actually stopped. All the resources are released properly.

From the application operations perspective, the above definition looks good.

But, If ConcurrentContainer `running` status has to be set to `false` only when all the containers are stopped. Then it looks as below.

1) set `running` true if all the contianers are running. `running` status true,  indicates `green` and All is Well.
2) 'running` state remains `true` even if one or more containers stopped. `running` status true, indicates `green` and All is Well. But, not correct!!

App dev always has to iterate over the containers to track the app condition. `running` status is not useful.

Reply:
The problem with `false` just for one child container that the rest are still running and we don't know about that from parent container since, as you said, it has to return `false`.
If that is what you want to implement here, then it is not OK to allow to stop any child container individually.
If we allow (and I don't see why not), then `false` for parent container would lead to the resource leak in the next `start()` call. 
Pay attention that we don't stop any running containers over there, but just create new instances.
Plus pay attention that `this.startedContainers.set(0);` is wrong here, because we don't take into account those running containers at the moment.
And all of that just because we decided to return `false` for the situation when at least one child container is stopped.

Reply:
> The problem with `false` just for one child container that the rest are still running and we don't know about that from parent container since, as you said, it has to return `false`.

If it returns `false`, indicates one, more or all containers are stopped. We can know, what containers are running by iterating over the containers.

Reply:
> If that is what you want to implement here, then it is not OK to allow to stop any child container individually.
> If we allow (and I don't see why not), then `false` for parent container would lead to the resource leak in the next `start()` call.

As per the existing logic running containers are stopped, if `stop` is called on `ConcurrentContainer`. There is no resource leak here. Please suggest here.

Reply:
> Plus pay attention that `this.startedContainers.set(0);` is wrong here, because we don't take into account those running containers at the moment.

this.startedContainers.set(0) -- Indicates initially no child container has started. Once started, it will be incremented and gets decremented once stopped.

Reply:
> The problem with `false` just for one child container that the rest are still running and we don't know about that from parent container since, as you said, it has to return `false`.

As mentioned earlier, I am good to implement as per your comments.

set `running` status to `false` only if all the containers. otherwise, set to `true`

 I will verify one more time and confirm this. I thought  of putting my views. Looking for your final review of my earlier comments before I implement as you mentioned.